### ISSUE TITLE
Allow returning `HITMAXBUTTON` from `WM_NCHITTEST` for Windows snap layout support

### ISSUE THREAD

**Issue by sphw** [rust-windowing/winit#3884]
### Description

In Windows 11, when you hover over the maximize button a window is displayed that shows various layouts you can "snap" the window to. When drawing a custom titlebar, these options don't work unless you return `HTMAXBUTTON` from the `WM_NCHITTEST` handler.

As far as I can tell this isn't possible in winit at the moment, and best I can figure there isn't a work around available. 

### Relevant platforms

Windows

**Comment by notgull** [comment]
> When drawing a custom titlebar, these options don't work unless you return `HTMAXBUTTON` from the `WM_NCHITTEST` handler.

I don't think there is something similar for any other platform, so it's probably outside the scope of `winit`. In addition it should be achievable through the [`with_msg_hook`](https://docs.rs/winit/latest/x86_64-pc-windows-msvc/winit/platform/windows/trait.EventLoopBuilderExtWindows.html#tymethod.with_msg_hook) function.

**Comment by sphw** [comment]
> I don't think there is something similar for any other platform, so it's probably outside the scope of `winit`. In addition it should be achievable through the [`with_msg_hook`](https://docs.rs/winit/latest/x86_64-pc-windows-msvc/winit/platform/windows/trait.EventLoopBuilderExtWindows.html#tymethod.with_msg_hook) function.

Fair enough on this being platform specific, but it is a very common issue for people trying to make fully featured desktop windows apps. 

I might be missing something, but [`with_msg_hook`](https://docs.rs/winit/latest/x86_64-pc-windows-msvc/winit/platform/windows/trait.EventLoopBuilderExtWindows.html#tymethod.with_msg_hook) seems to return `bool` that either lets winit handle the event or not. To correctly handle WM_NCHITTEST, you need to return a value from the `lpfnWndProc` callback. As far as I can tell you can't control the return value of `lpfnWndProc`
 
In my mind this would be similar to [`show_window_menu`](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.show_window_menu) in that it allows developers to create a more native feeling experience. I would be happy for it to be an additional method on `EventLoopBuilderExtWindows` or something similar to [`drag_resize_window`](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.drag_resize_window) on `Window`

**Comment by DataTriny** [comment]
Your only option @sphw at the moment is to use window subclassing. We use it in AccessKit to return handles to COM interfaces in response to the `WM_GETOBJECT` message. You can see how we implement it [here](https://github.com/AccessKit/accesskit/blob/main/platforms/windows/src/subclass.rs).

I hope that API changes to winit 0.31 will allow us to get rid of this hack though.

**Comment by sphw** [comment]
Thanks for the suggestion! I'll give it a try

**Comment by notgull** [comment]
Yeah subclassing may be the best option here, since this is effectively "add a new function to a window handle". I would definitely be open to making subclassing easier, either through `winit` or through another library dedicated to this kind of subclassing.

**Comment by Zbrooklyn** [comment]
One caveat worth recording here, since @DataTriny's subclassing suggestion is the accepted answer on this thread and it holds for a plain winit window — but it **cannot** work for anyone who has a child window covering the client area, which includes every webview-based app (Tauri/wry, and this is the same reason the equivalent Tauri issue [#4531](https://github.com/tauri-apps/tauri/issues/4531) has stayed open for four years).

`WM_NCHITTEST` is delivered to whichever window is under the cursor. With a webview, that's the Chromium child (`Chrome_RenderWidgetHostHWND`), not your winit window — so a subclass proc on the top-level HWND never sees the message at all. Subclassing the descendants doesn't rescue it either: Chromium uses its own internal input routing, and forcibly swapping its WndProc breaks rendering.

What does work at the application level is a separate small transparent child window positioned over the maximize button, owning `WM_NCHITTEST` for just that rectangle and returning `HTMAXBUTTON`. Style set matters — `WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | WS_OVERLAPPED`, **no** extended styles, `NULL_BRUSH`. Invisibility has to come from never painting; `WS_EX_LAYERED` costs you the hit test, and `WS_EX_TRANSPARENT` makes it hit-test-transparent, which defeats the entire purpose.

Because it owns the mouse in that rectangle, it also has to forward hover and click back to the app, otherwise the button goes visually dead.

Full write-up of the failure modes, plus two runnable templates and a verification script that asserts against a live window: https://github.com/Zbrooklyn/tauri-snap-layouts

Not arguing this belongs in winit — @notgull's scoping call seems right to me. Just leaving the caveat so the next person who follows the subclassing advice into a webview app knows why it silently does nothing.
